code/controllers/subsystem/demo.dm

```diff
@@ -0,0 +1,427 @@
+SUBSYSTEM_DEF(demo)
+	name = "Demo"
+	wait = 1
+	flags = SS_TICKER
+	init_order = INIT_ORDER_DEMO
+	runlevels = RUNLEVELS_DEFAULT | RUNLEVEL_LOBBY
+
+	var/list/pre_init_lines = list() // stuff like chat before the init
+	var/list/icon_cache = list()
+	var/list/icon_state_caches = list()
+	var/list/name_cache = list()
+
+	var/list/marked_dirty = list()
+	var/list/marked_new = list()
+	var/list/marked_turfs = list()
+	var/list/del_list = list()
+
+	var/last_written_time = null
+	var/last_chat_message = null
+
+	// stats stuff
+	var/last_queued = 0
+	var/last_completed = 0
+
+/datum/controller/subsystem/demo/proc/write_time()
+	var/new_time = world.time
+	if(last_written_time != new_time)
+		if(initialized)
+			WRITE_LOG_NO_FORMAT(GLOB.demo_log, "time [new_time]\n")
+		else
+			pre_init_lines += "time [new_time]"
+	last_written_time = new_time
+
+/datum/controller/subsystem/demo/proc/write_event_line(line)
+	write_time()
+	if(initialized)
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, "[line]\n")
+	else
+		pre_init_lines += line
+
+/datum/controller/subsystem/demo/proc/write_chat(target, text)
+	var/target_text = ""
+	if(target == GLOB.clients)
+		target_text = "world"
+	else if(islist(target))
+		var/list/target_keys = list()
+		for(var/T in target)
+			var/client/C = CLIENT_FROM_VAR(T)
+			if(C)
+				target_keys += C.ckey
+		if(!target_keys.len)
+			return
+		target_text = jointext(target_keys, ",")
+	else
+		var/client/C = CLIENT_FROM_VAR(target)
+		if(C)
+			target_text = C.ckey
+		else
+			return
+	write_event_line("chat [target_text] [last_chat_message == text ? "=" : json_encode(text)]")
+	last_chat_message = text
+
+/datum/controller/subsystem/demo/Initialize()
+	WRITE_LOG_NO_FORMAT(GLOB.demo_log, "demo version 1\n") // increment this if you change the format
```

Comment:
The version should probably be a define somewhere for clairity. Also, can we get a config option to enable this whole thing?